core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala

```diff
@@ -997,173 +996,45 @@ object GroupMetadataManager {
-    0 -> GROUP_METADATA_VALUE_SCHEMA_V0,
-    1 -> GROUP_METADATA_VALUE_SCHEMA_V1,
-    2 -> GROUP_METADATA_VALUE_SCHEMA_V2,
-    3 -> GROUP_METADATA_VALUE_SCHEMA_V3)
-
-  private val CURRENT_OFFSET_KEY_SCHEMA = schemaForKey(CURRENT_OFFSET_KEY_SCHEMA_VERSION)
-  private val CURRENT_GROUP_KEY_SCHEMA = schemaForKey(CURRENT_GROUP_KEY_SCHEMA_VERSION)
-  private val CURRENT_GROUP_METADATA_VALUE_SCHEMA_VERSION = GROUP_VALUE_SCHEMAS.keySet.max
-
-  private def schemaForKey(version: Int) = {
-    val schemaOpt = MESSAGE_TYPE_SCHEMAS.get(version)
-    schemaOpt match {
-      case Some(schema) => schema
-      case _ => throw new KafkaException("Unknown message key schema version " + version)
-    }
-  }
-
-  private def schemaForOffsetValue(version: Int) = {
-    val schemaOpt = OFFSET_VALUE_SCHEMAS.get(version)
-    schemaOpt match {
-      case Some(schema) => schema
-      case _ => throw new KafkaException("Unknown offset schema version " + version)
+  /**
+   *
+   * Statically check to make sure that the generated code always stays in sync with the overall protocol
+   * @param clz generated class
+   * @param actual actual version
+   * @param expected expected version
+   * @return correct version
+   */
+  private def checkVersionOfGeneratedCode(clz: Class[_], actual: Short, expected: Short): Short = {
+    if (actual != expected) {
+      throw new IllegalArgumentException(
+        s"core/src/main/resources/common/message/${clz.getSimpleName}.json needs to be updated to match the " +
+          s"latest assignment protocol version. ${clz.getSimpleName} only supports up to  ["
+          + actual + "] but needs to support up to [" + expected + "].");
     }
+    expected
   }
 
-  private def schemaForGroupValue(version: Int) = {
-    val schemaOpt = GROUP_VALUE_SCHEMAS.get(version)
-    schemaOpt match {
-      case Some(schema) => schema
-      case _ => throw new KafkaException("Unknown group metadata version " + version)
-    }
+  private val CURRENT_OFFSET_KEY_SCHEMA_VERSION = checkVersionOfGeneratedCode(
+    clz = classOf[GenOffsetCommitRequest],
+    actual = (GenOffsetCommitRequest.SCHEMAS.length - 1).toShort,
+    expected = 1.toShort)
+  private val CURRENT_GROUP_KEY_SCHEMA_VERSION = checkVersionOfGeneratedCode(
+    clz = classOf[GroupMetadataRequest],
+    actual = (GroupMetadataRequest.SCHEMAS.length - 1).toShort,
+    expected = 2.toShort)
+  private val CURRENT_GROUP_METADATA_VALUE_SCHEMA_VERSION = checkVersionOfGeneratedCode(
+    clz = classOf[GroupMetadataResponse],
+    actual = (GroupMetadataResponse.SCHEMAS.length - 1).toShort,
+    expected = 3.toShort)
+
+  private def serializeMessage(version: Short, message: Message): Array[Byte] = {
+    val cache = new ObjectSerializationCache()
+    val size = message.size(cache, version)
+    val bytes = ByteBuffer.allocate(2 + size)
+    val accessor = new ByteBufferAccessor(bytes)
+    accessor.writeShort(version)
+    message.write(accessor, cache, version)
+    bytes.array()
```

Comment:
We have merged my PR that contains this piece of code as well: https://github.com/apache/kafka/commit/466f8fd21c6651ea5daa50154239e85fa629dbb4#diff-bad29ccb1aba700e1badeff62f1a86b7R178. I think that we should find a common place where we could put that piece in order to avoid having it twice.

`org.apache.kafka.common.protocol.MessageUtil` may be a good place. What do you think?